### ISSUE TITLE
Poor performance when simulation has a lot of pairs captured in stateful mode

### ISSUE THREAD

**Issue by aosavitski** [SpectoLabs/hoverfly#855]
Simulations captured with stateful mode are very slow, because copy of the state.State map is done in /hoverfly/core/matching/state_matcher.go and /hoverfly/core/matching/strongest_match_strategy_test.go

`state_matcher.go`
`currentState.RWMutex.RLock()`
`copyState := util.CopyMap(currentState.State)`
`currentState.RWMutex.RUnlock()`

`strongest_match_strategy.go`
`state.RWMutex.RLock()`
`copyState := util.CopyMap(state.State)`
`state.RWMutex.RUnlock()`

And problem is not in locking added in [hoverfly#767](https://github.com/SpectoLabs/hoverfly/issues/767), but in coping this map in the loop. In state_matcher.go we are interested in just one pair from this map, but we are coping whole map for each pair... In strongest_match_strategy.go we put this copy in the ClosestMiss for each miss with the same or better miss score. Why? Just to show this whole map in the log at the end? I think it could be improved. As a quick an dirty fix I can offer this solution:
We are doing this copyMap only once in matching_strategy_runner.go, before we start loop. And then StateMatcher and PostMatching will have as argument `copyState map[string]string` instead of `currentState *state.State`.

**Comment by aosavitski** [comment]
> @tommysitu 
> @aosavitski maybe the locking and map copy can be replaced by syncmap >https://godoc.org/golang.org/x/sync/syncmap. Can you try it out please? You’ve got the perfect >environment to test out this fix. A PR will be really welcome.

@tommysitu [documentation](https://tip.golang.org/pkg/sync/#Map) saying: `The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination`
state.State is already an API on top of original map with synchronisation. So may be better is not to replace it, but just stop accessing to the map data directly?
So in state_matcher.go instead of creating a copy and accessing value with copyMap[key] it could be:
```
currentStateValue, ok := currentState.GetState(key)
if !ok {
	matched = false
}
if currentStateValue != value {
	matched = false
} else {
	score++
}
```
instead of current:
```
currentState.RWMutex.RLock()
copyState := util.CopyMap(currentState.State)
currentState.RWMutex.RUnlock()
for key, value := range requiredState {
if _, ok := copyState[key]; !ok {
	matched = false
}
if copyState[key] != value {
	matched = false
} else {
	score++
}
```
and in ClosestMiss struct `State          map[string]string` could be replaced with `State          *state.State`
BUT at the moment when we will print state in ClosestMiss.GetMessage(), it could be modified already by other calls. And sync.Map will not help here too. So, probably the solution should be a combination - in state_matcher.go data in the map should not be accessed directly, only via API of state.State. And in strongest_match_strategy.go copy of the map will be used, but it will be created once outside the loop

**Comment by aosavitski** [comment]
> A PR will be really welcome.

If you agree with such a solution, I can do a PR, no problem. But then first I need to find out how to run tests:) Because I need to modify them. I am new to golang.

**Comment by tommysitu** [comment]
@aosavitski  I suggest looking into sync.map is because it doesn't lock the whole map while getting a key, so it could reduce lock contention when comparing with an array of the requiredState. But the main problem as you pointed out is the copy of map in a loop. So I think what you suggest should fix the problem. 

You just need to do `make test` to run all the tests. Don't worry if some middleware tests fail as they need some setup to work. Trying out the build with your simulation file would be good.

**Comment by tommysitu** [comment]
Thank you for your PR, this is now released in [v1.1.1](https://github.com/SpectoLabs/hoverfly/releases/tag/v1.1.1)
